### PR TITLE
Update Kokkos library in LAMMPS to v3.1

### PR DESCRIPTION
**Pull request by stanmoore1**
**Summary**

Update Kokkos library in LAMMPS to v3.1

**Related Issues**

None.

**Author(s)**

Stan Moore (Sandia)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No issues.

### CODE REVIEW

**Comment by akohlmey** [comment]
@stanmoore1 does this still need the patch to work? Otherwise we should remove the code to transparently download it from the KOKKOS.cmake file.
Beyond that, can you comment on which additional Kokkos options should be documented in the manual. While it may be tricky to apply them through the "magic" Makefile, it will be straightforward via CMake.
We already automatically enforce CUDA lambdas, so that is not needed, but there is the Kokkos_CUDA_UVM option or Kokkos_ENABLE_HWLOC and Kokkos_ENABLE_LIBNUMA (are they of use with LAMMPS, anyway?) and a few more. It would be nice to have those options listed and described, that are of use with LAMMPS. Now that we have singularity containers for ROCm and CUDA and both, it would also be easy to test compilation with "HIP", so that could be documented, too.

**Comment by stanmoore1** [comment]
Most of the patches are included. There is a small (persistent) patch needed to `Makefile.kokkos`. At this point we could remove the `lib/kokkos` directory and use `Install.py`. I'm a little hesitant though since it could break people's workflow and makes it harder to build.

**Comment by akohlmey** [comment]
> Most of the patches are included. There is a small (persistent) patch needed to `Makefile.kokkos`. At this point we could remove the `lib/kokkos` directory and use `Install.py`. I'm a little hesitant though since it could break people's workflow and makes it harder to build.

Ok, then I will remove the application of the patch from the CMake script, as that does not use Makefile.kokkos.
I suggest we leave the decision about unbundling the Kokkos library until after our next in-person meeting, or make it a topic on a future video con-call (a related topic in need of a discussion, is the growing size of the LAMMPS source archive and whether to break it down and then how, so that could be added then). For now we better focus on mending what got broken with the big changes in the last patch release and what will get broken with the couple more big changes planned for the next patch, and then wait with any changes with the potential to break things until after the next stable release.

**Comment by stanmoore1** [comment]
>Ok, then I will remove the application of the patch from the CMake script, as that does not use Makefile.kokkos.
>I suggest we leave the decision about unbundling the Kokkos library until after our next in-person meeting, or make it a topic on a future video con-cal

Sounds good. I'll work on the docs. 

>Kokkos_CUDA_UVM option or Kokkos_ENABLE_HWLOC and Kokkos_ENABLE_LIBNUMA

`Kokkos_CUDA_UVM ` can be useful. I've never used the other options.

**Comment by akohlmey** [comment]
> Sounds good. I'll work on the docs.

Perhaps we should integrate the section `Advanced Kokkos options` from `doc/src/Speed_packages.rst` into `doc/src/Build_extras.rst`. If you do the conventional make part, I can "translate" it for CMake.
 
> > Kokkos_CUDA_UVM option or Kokkos_ENABLE_HWLOC and Kokkos_ENABLE_LIBNUMA
> 
> `Kokkos_CUDA_UVM ` can be useful. I've never used the other options.

hwloc is mentioned in `doc/src/Run_options.rst` and `doc/src/Speed_packages.rst`, though, so we should mention how to enable/disable it also for CMake (ideal would be to detect and enable one of those if available) and I suspect that libnuma support would do the equivalent just using a different library.

I did not see a 3.1 release on GitHub yet, so I cannot update `cmake/Modules/Packages/KOKKOS.cmake` at the moment.

**Comment by stanmoore1** [comment]
>I did not see a 3.1 release on GitHub yet, so I cannot update cmake/Modules/Packages/KOKKOS.cmake at the moment.

It was just posted a moment ago.

**Comment by akohlmey** [comment]
@junghans just tested this branch with `-DDOWNLOAD_KOKKOS` apparently the various `-DKokkos_XXX` options need to be forwarded from the lammps target to building the downloaded library (or else it will always build for serial and no CUDA). do you have a suggestion for how to add this? essentially all variables that start with `Kokkos_` need to be passed to building the Kokkos lib.

**Comment by stanmoore1** [comment]
Not sure, I wonder what other projects that use Kokkos do.

**Comment by akohlmey** [comment]
> Not sure, I wonder what other projects that use Kokkos do.

i think i found something after a session of googling and try-n-error.

**Comment by akohlmey** [comment]
this seems to be working. curious what @junghans will think/suggest
this currently is always using the nvcc_wrapper from the bundled library (that is what the preset sets), but for as long as we keep bundling kokkos this is not a concern. and we can figure out how to fix it later when that day comes.

**Comment by junghans** [comment]
> @junghans just tested this branch with `-DDOWNLOAD_KOKKOS` apparently the various `-DKokkos_XXX` options need to be forwarded from the lammps target to building the downloaded library (or else it will always build for serial and no CUDA). do you have a suggestion for how to add this? essentially all variables that start with `Kokkos_` need to be passed to building the Kokkos lib.

I would do a loop:
```
foreach(_OPT ${KOKKOS_ALLOWED_OPTION})
  if(DEFINED ${_OPT})
    list(APPEND KOKKOS_EXTRA_CMAKE_OPTS "-D${_OPT}=${${_OPT}}")
  endif()
endforeach()
```
and then use `${KOKKOS_EXTRA_CMAKE_OPTS}` in `CMAKE_ARGS`.

Alternatively we could make an `KOKKOS_EXTRA_CMAKE_OPTS` option an leave it to the user.

**Comment by junghans** [comment]
> this seems to be working. curious what @junghans will think/suggest
> this currently is always using the nvcc_wrapper from the bundled library (that is what the preset sets), but for as long as we keep bundling kokkos this is not a concern. and we can figure out how to fix it later when that day comes.

Great work that is pretty much, what I suggested without seeing your commit ;-)

**Comment by akohlmey** [comment]
@stanmoore1 just pushed some changes to this PR that update the documentation to match what is provided by Kokkos 3.1 and also moved the "advanced" build options from the Speed_kokkos.rst page to the Build_extras.rst page. Please have a look and give me some notes, if you would like me to make some more changes or made some mistakes. Tried building for AMD GPUs with HIP, but so far no luck, therefore I didn't document it (yet).

**Comment by stanmoore1** [comment]
Thanks @akohlmey. FYI, nearly all of the regression tests are passing, with only a couple I need to look into.

File: lib/kokkos/BUILD.md [hunk 1]
```diff
@@ -40,6 +40,13 @@ cmake ${srcdir} \
 ````
 which activates the OpenMP backend. All of the options controlling device backends, options, architectures, and third-party libraries (TPLs) are given below.
 
+## Platform-specific Problems
+
+### Cray
+
+* The Cray compiler wrappers do static linking by default. This seems to break the Kokkos build. You will likely need to set the environment variable `CRAYPE_LINK_TYPE=dynamic` in order to link correctly. Kokkos warns during configure if this is missing.
```
Comment (stanmoore1) [inline lib/kokkos/BUILD.md:47]: @akohlmey Is this referring to the `dlopen` warning? If so I think that can be safely ignored.

**Comment by stanmoore1** [comment]
I'm down to only one failing CUDA regression test.

File: lib/kokkos/BUILD.md (hunk 1 above)
Comment (akohlmey) [reply lib/kokkos/BUILD.md:47]: no idea. that is a question for the kokkos folks. i don't have an account on a cray box currently.

**Comment by stanmoore1** [comment]
That last failing test is now passing.

**Pull request merged**